internal/services/fingerprint/consumer.go

```diff
@@ -0,0 +1,170 @@
+package fingerprint
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/DIMO-Network/device-data-api/internal/config"
+	"github.com/DIMO-Network/device-data-api/internal/helpers"
+	"github.com/DIMO-Network/device-data-api/models"
+	deviceapi "github.com/DIMO-Network/devices-api/models"
+	"github.com/DIMO-Network/shared/kafka"
+	"github.com/volatiletech/sqlboiler/v4/boil"
+	"github.com/volatiletech/sqlboiler/v4/queries/qm"
+
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/DIMO-Network/device-data-api/internal/services/issuer"
+	"github.com/DIMO-Network/shared"
+	"github.com/DIMO-Network/shared/db"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+)
+
+type Event struct {
+	shared.CloudEvent[json.RawMessage]
+	Signature string `json:"signature"`
+}
+
+type Consumer struct {
+	logger *zerolog.Logger
+	iss    *issuer.Issuer
+	DBS    db.Store
+}
+
+func NewConsumer(dbs db.Store, iss *issuer.Issuer, log *zerolog.Logger) *Consumer {
+	return &Consumer{
+		DBS:    dbs,
+		logger: log,
+		iss:    iss,
+	}
+}
+
+func RunConsumer(ctx context.Context, settings *config.Settings, logger *zerolog.Logger, i *issuer.Issuer, dbs db.Store) error {
+	consumer := NewConsumer(dbs, i, logger)
+
+	if err := kafka.Consume(ctx, kafka.Config{
+		Brokers: strings.Split(settings.KafkaBrokers, ","),
+		Topic:   settings.DeviceFingerprintTopic,
+		Group:   settings.DeviceFingerprintConsumerGroup,
+	}, consumer.HandleDeviceFingerprint, logger); err != nil {
+		logger.Fatal().Err(err).Msg("couldn't start device fingerprint consumer")
+	}
+
+	logger.Info().Msg("Starting transaction request status listener.")
+
+	return nil
+}
+
+// defaultCredDuration is the default lifetime for VIN credentials. It's meant to cover the
+// current reward week, but contains an extra day for safety.
+var defaultCredDuration = 8 * 24 * time.Hour
+var userDeviceDataPrimaryKeyColumns = []string{models.UserDeviceDatumColumns.UserDeviceID, models.UserDeviceDatumColumns.IntegrationID}
+
+func (c *Consumer) HandleDeviceFingerprint(ctx context.Context, event *Event) error {
+	if !common.IsHexAddress(event.Subject) {
+		return fmt.Errorf("subject %q not a valid address", event.Subject)
+	}
+	addr := common.HexToAddress(event.Subject)
+	signature := common.FromHex(event.Signature)
+	hash := crypto.Keccak256Hash(event.Data)
+
+	if recAddr, err := helpers.Ecrecover(hash.Bytes(), signature); err != nil {
+		return fmt.Errorf("failed to recover an address: %w", err)
+	} else if recAddr != addr {
+		return fmt.Errorf("recovered wrong address %s", recAddr)
+	}
+
+	VIN, err := ExtractVIN(event.Data)
+	if err != nil {
+		if err == ErrNoVIN {
+			return nil
+		}
+		return fmt.Errorf("couldn't extract VIN: %w", err)
+	}
+
+	ad, err := deviceapi.AftermarketDevices(
```

Comment:
this looks like a call to devices api database, not sure that will work here. There should be a grpc call to get user device by eth address.